server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1868,7 +1906,7 @@ public void prepareForIndexRecovery() {
             throw new IndexShardNotRecoveringException(shardId, state);
         }
         recoveryState.setStage(RecoveryState.Stage.INDEX);
-        assert currentEngineReference.get() == null;
+        assert this.currentEngine == null;
```

Comment:
I wonder if there are visibility guarantees to ensure that the assertion works?

Reply:
Indeed, we should hold the read lock while asserting.

Let me know if you want me to do it or if you do the change yourself.

Reply:
I'll take care 👍